### PR TITLE
Nix: remove HiDPI patches

### PR DESCRIPTION
**Pull request by fufexan**
#### Describe your PR, what does it fix/add?
- remove `hidpiXWayland` option
- rename `nvidiaPatches` -> `enableNvidiaPatches` for consistency with `enableXWayland`
- Complements https://github.com/hyprwm/hyprland-wiki/pull/282.

#### Is there anything you want to mention? (unchecked code, possible bugs, found problems, breaking compatibility, etc.)
Will need to be announced to the community when merged.

#### Is it ready for merging, or does it need work?
Ready.

### CODE REVIEW

**Comment by vaxerski** [comment]
reqs a wiki update ofc ass well

**Comment by q234rty** [comment]
> they are deprecated because xwayland:force_zero_scaling does exactly the same with less hacks

Unfortunately, I believe `xwayland:force_zero_scaling` as it currently is does not support multi monitor with different scale factors, since X11 does not support them except maybe with Qt.

**Comment by MrDuartePT** [comment]
@fufexan since you seems to be the maintainer of the Nix package for Hyprland you can give a look at this PR: https://github.com/hyprwm/Hyprland/pull/2874.

Doesn't seems to break the Nix package but I don't use Nix that why I not sure

**Comment by fufexan** [comment]
@q234rty as far as I know, the patches don't solve this issue either.

**Comment by q234rty** [comment]
@fufexan These patches partially mitigate this issue by letting the compositor do per-monitor downscaling, e.g. if one monitor is under 200% and another monitor is under 100%, Xwayland apps will see two 200% screens and the compositor will do the downscaling on the 100% monitor. This is of course not perfect but still much better than having very large Xwayland apps on the 100% monitor.

**Comment by fufexan** [comment]
@q234rty if the users desire that functionality, they can copy the code previous to these commits and override hyprland themselves. In the meantime, we will be waiting for an official solution.

**Comment by q234rty** [comment]
Whatever, I will just continue maintaining my AUR hyprland-hidpi-xprop-git package with the old approach :)

**Comment by vaxerski** [comment]
we are removing the patches because of my pressure to do so. 

The approach is shit and buggy, requires patching which already is a pain, and xwayland:force_zero_scaling does exactly the same.

**Comment by q234rty** [comment]
> and xwayland:force_zero_scaling does exactly the same.

As mentioned above, `xwayland:force_zero_scaling` does not handle multi monitor with different scale factors, so it does not "do exactly the same"

**Comment by vaxerski** [comment]
does it not? how so? it _removes_ scaling.

**Comment by q234rty** [comment]
Yes, it removes scaling by the compositor, but no, X11 itself does not handle multi monitor with different scale factors, so letting X11 see the same scale factor across all monitors and letting the compositor do the downscaling is helpful, which is what the old approach did.

File: nix/hm-module.nix [hunk 1]
```diff
@@ -7,8 +7,7 @@ self: {
   cfg = config.wayland.windowManager.hyprland;
   defaultHyprlandPackage = self.packages.${pkgs.stdenv.hostPlatform.system}.default.override {
     enableXWayland = cfg.xwayland.enable;
-    hidpiXWayland = cfg.xwayland.hidpi;
-    inherit (cfg) nvidiaPatches;
+    inherit (cfg) enableNvidiaPatches;
```
[the remark this answers is not in the document]
Comment (spikespaz) [reply nix/hm-module.nix:10]: This is out of scope for this PR:
I have handled the package overrides a little differently. The way I did it allows the user to set the `package` but also use the `enable*` options that override it.
See: <https://github.com/spikespaz/hyprland-flake/blob/3642290700deb55b097275d1425c63327431dd4c/hm-module/config.nix#L234-L239>

File: nix/module.nix [hunk 2]
```diff
@@ -30,47 +27,32 @@ in {
...
-        hyprland.packages.''${pkgs.stdenv.hostPlatform.system}.default.override {
-          enableXWayland = config.programs.hyprland.xwayland.enable;
-          hidpiXWayland = config.programs.hyprland.xwayland.hidpi;
-          inherit (config.programs.hyprland) nvidiaPatches;
-        }
-      '';
-      example = literalExpression "pkgs.hyprland";
+    package = mkPackageOptionMD inputs.packages.${pkgs.system} "hyprland" {};
+
+    finalPackage = mkOption {
```
Comment (spikespaz) [inline nix/module.nix:32]: See my other comment for how I set up the `package'`, but as this currently currently stands, you're missing `hidden = true`. Unless you intend to make the `finalPackage` available through `config` for the sake of recursively referencing the binaries in configuration.

File: nix/module.nix [hunk 3]
```diff
@@ -30,47 +27,32 @@ in {
...
-      hidpi =
-        mkEnableOption null
-        // {
-          description = mdDoc ''
-            Enable HiDPI XWayland, based on [XWayland MR 733](https://gitlab.freedesktop.org/xorg/xserver/-/merge_requests/733).
-            See <https://wiki.hyprland.org/Nix/Options-Overrides/#xwayland-hidpi> for more info.
-          '';
-        };
-    };
+    portalPackage = mkPackageOptionMD inputs.xdph.packages.${pkgs.system} "xdg-desktop-portal-hyprland" {};
```
Comment (spikespaz) [inline nix/module.nix:47]: If XDPH may receive its own options in the future, consider `portal.package`. This may seem to conflict with my earlier suggestion pertaining to `xwayland.enable`, but `xwayland` no longer has any other options.

File: nix/module.nix [hunk 4]
```diff
@@ -30,47 +27,32 @@ in {
...
+
+    enableNvidiaPatches = mkEnableOption (mdDoc "patching wlroots for better Nvidia support");
   };
 
   config = mkIf cfg.enable {
-    environment = {
-      systemPackages = [cfg.package];
-
-      sessionVariables = {
-        NIXOS_OZONE_WL = mkDefault "1";
```
Comment (spikespaz) [inline nix/module.nix:71]: Why was this removed? Is it no longer relevant for some reason?

File: nix/hm-module.nix [hunk 5]
```diff
@@ -86,19 +84,9 @@ in {
...
-      hidpi =
-        lib.mkEnableOption null
-        // {
-          description = lib.mdDoc ''
-            Enable HiDPI XWayland, based on [XWayland MR 733](https://gitlab.freedesktop.org/xorg/xserver/-/merge_requests/733).
-            See <https://wiki.hyprland.org/Nix/Options-Overrides/#xwayland-hidpi> for more info.
-          '';
-        };
-    };
+    xwayland.enable = lib.mkEnableOption (lib.mdDoc "XWayland") // {default = true;};
```
[the remark this answers is not in the document]
Comment (spikespaz) [reply nix/hm-module.nix:87]: https://github.com/NixOS/nixpkgs/blob/nixos-23.05/nixos/modules/programs/xwayland.nix#L15
If this is helpful, ignore my suggestion on renaming this option.

File: nix/wlroots.nix [hunk 6]
```diff
@@ -1,64 +1,32 @@
...
-  fetchpatch,
-  lib,
   hwdata,
-  libliftoff,
   libdisplay-info,
-  hidpiXWayland ? true,
-  enableXWayland ? true,
-  nvidiaPatches ? false,
+  libliftoff,
+  enableNvidiaPatches ? false,
```
Comment (spikespaz) [inline nix/wlroots.nix:9]: `withNvidiaPatches`?

File: nix/wlroots.nix [hunk 7]
```diff
@@ -1,64 +1,32 @@
...
-          })
-        ])
-        ++ (lib.optionals nvidiaPatches [
-          ./patches/wlroots-nvidia.patch
-        ]);
-      buildInputs = old.buildInputs ++ [hwdata libliftoff libdisplay-info];
+wlroots.overrideAttrs (old: {
+  inherit version src;
+
+  pname =
```
Comment (spikespaz) [inline nix/wlroots.nix:14]: Because of Alejandra, I prefer: `"${old.pname}-hyprland${lib.optionalString enableNvidiaPatches "-nvidia"}"`

**Review by spikespaz — COMMENTED** [review]
I left a many comments, most are nitpicks or questions. This looks good as it is. There are some suggestions that do not align with the description of the PR.

File: nix/default.nix [hunk 8]
```diff
@@ -26,114 +26,113 @@
   xcbutilwm,
   xwayland,
   debug ? false,
+  enableNvidiaPatches ? false,
```
[the remark this answers is not in the document]
Comment (fufexan) [reply nix/default.nix:29]: The Nixpkgs PR included the same change and it got merged, so I won't rename this now.

File: nix/default.nix [hunk 9]
```diff
@@ -26,114 +26,113 @@
...
-          libdrm
-          libinput
-          libxkbcommon
-          mesa
-          pango
-          udis86
-          wayland
-          wayland-protocols
-          pciutils
-          (wlroots.override {inherit enableXWayland hidpiXWayland nvidiaPatches;})
```
[the remark this answers is not in the document]
Comment (fufexan) [reply nix/default.nix:86]: Because it was never used apart from an assertion for `hidpiXWayland`.

File: nix/hm-module.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (fufexan) [reply nix/hm-module.nix:10]: I feel the difference is minimal.

File: nix/module.nix (hunk 2 above)
Comment (fufexan) [reply nix/module.nix:32]: That's the usecase, indeed.

File: nix/module.nix (hunk 3 above)
Comment (fufexan) [reply nix/module.nix:47]: We'll be adding that if it is the case. Currently, no such options are planned. Also I'm not aware of any portal that can be configured in any way.

File: nix/module.nix (hunk 4 above)
Comment (fufexan) [reply nix/module.nix:71]: I mirrored the Nixpkgs Hyprland module. Might as well just get rid of this one and set the proper packages for the Nixpkgs one.

But then again, NixOS Stable users will complain that module brokey or something. ugh.

File: nix/wlroots.nix (hunk 7 above)
Comment (fufexan) [reply nix/wlroots.nix:14]: Hopefully nixfmt gets blessed as *the* formatter soon, so we won't have to worry.
In the meantime, I'll change it.

File: nix/hm-module.nix (hunk 5 above)
[the remark this answers is not in the document]
Comment (fufexan) [reply nix/hm-module.nix:87]: Changed the doc string.

File: nix/hm-module.nix [hunk 10]
```diff
@@ -86,19 +84,9 @@ in {
...
 
-    nvidiaPatches = lib.mkEnableOption (lib.mdDoc "patching wlroots for better Nvidia support.");
+    enableNvidiaPatches = lib.mkEnableOption (lib.mdDoc "patching wlroots for better Nvidia support.");
```
[the remark this answers is not in the document]
Comment (fufexan) [reply nix/hm-module.nix:89]: Done.

File: nix/default.nix (hunk 9 above)
[the remark this answers is not in the document]
Comment (fufexan) [reply nix/default.nix:86]: I've added it back and propagated it to the package.

[the remark this answers is not in the document]
Comment (spikespaz) [reply nix/default.nix:86]: If it isn't necessary, and you removed it because of that, why did you add it back in? :confused:

File: nix/hm-module.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (spikespaz) [reply nix/hm-module.nix:10]: > I feel the difference is minimal.

True, but it still depends. The only real difference is if someone wants to override the package but still have the module options applied, your way would not work and would mean that the package takes precedence over options. From your perspective though, if they override the package, they probably know what they're doing and would supply those arguments in the override anyway.

I just did this so that you could use a fork or a different branch of Hyprland and have the options applied, assuming the user is not experienced and would be confused that the options stop working.

[the remark this answers is not in the document]
Comment (fufexan) [reply nix/hm-module.nix:10]: No, I literally mean just the formatting is different. We set `package` and the options get applied to it, and that forms `finalPackage` which we then use everywhere else. It's equivalent to your code.

[the remark this answers is not in the document]
Comment (fufexan) [reply nix/hm-module.nix:10]: Oh my bad, that's in the NixOS module (and the upstream HM module).

I didn't want to spend any more time working on this, provided we'd be integrating your module soon:tm:.

File: nix/module.nix (hunk 3 above)
Comment (spikespaz) [reply nix/module.nix:47]: In `hyprland.portal`, I do override `UseIn`: <https://github.com/hyprwm/xdg-desktop-portal-hyprland/blob/master/hyprland.portal#L4>

That could be exposed as an option. I do this: <https://github.com/spikespaz/dotfiles/blob/master/hm-modules/desktop-portals.nix>

There was recently a PR to `xdg-desktop-portal-wlr` that removed the necessity for this, something to do with precedence. I can't seem to find it though, but maybe I'm looking at the wrong thing.

This issue seems relevant: <https://github.com/emersion/xdg-desktop-portal-wlr/issues/268>

File: nix/wlroots.nix (hunk 7 above)
Comment (spikespaz) [reply nix/wlroots.nix]: > Hopefully nixfmt gets blessed as the formatter soon

You could help with that by using it :smile:

**Review by spikespaz — APPROVED** [review]

File: nix/default.nix (hunk 9 above)
[the remark this answers is not in the document]
Comment (fufexan) [reply nix/default.nix:86]: Well it was never applied properly, but now it is. I don't use Hyprland xwayland-free so I don't know whether this even affects anything.

File: nix/module.nix (hunk 3 above)
Comment (fufexan) [reply nix/module.nix:47]: Oh, I see. From past issues, it's more likely that the GNOME portal causes issues and would need that, and not Hyprland. I think it's best to wait until the new mechanism arrives.

File: nix/wlroots.nix (hunk 7 above)
Comment (fufexan) [reply nix/wlroots.nix]: Yes, too bad [the PR](https://github.com/serokell/nixfmt/pull/118) is still a draft.
